### ISSUE TITLE
Race condition in thread cancellation early after start

### ISSUE THREAD

**Issue by Werkov** [rsyslog/rsyslog#1959]
## Steps to reproduce

I reproduced this on SLES12 SP2, relevant packages versions:

> kernel-default-4.4.90-92.45.1.x86_64
> glibc-2.22-61.3.x86_64
> rsyslog-8.4.0-16.2.x86_64

It was QEMU/KVM with a single vCPU.

Based on default rsyslog configuration, I set up remote forwarding:
```
 cat /etc/rsyslog.d/remote-client.conf
 action(type="omfwd" target="10.100.52.220" port="514" protocol="udp"
         queue.filename="remote" queue.spoolDirectory="/var/log/rsyslog-spool/rsyslog"
         queue.type="LinkedList" queue.maxfilesize="1g" queue.saveonshutdown="on")
```

The issue is that occasionally rsyslog.service stop operation times out. Originally, this was observed only during shutdown but later I was able to reproduce it during any increased load.

Reproduction steps are thus:
```
# Terminal 1
./make-load.sh

# Terminal 2
systemctl stop rsyslog.service
```
The load was simply a looping bash:
```bash
cat make-load.sh
#!/bin/bash

N=${1:-17}
T=10000000

function worker() {
        i=0
        while true ; do
                i=$(($i+1))
                if [ $i -gt $T ] ; then
                        break
                fi
        done
}

for w in `seq $N` ; do
        worker &
done

wait
```

### Expected behavior

rsyslog daemon terminates eventually.

### Actual behavior

rsyslog daemon doesn't terminate and has to be `SIGKILL`ed.

## Analysis

With strace I was able to see that main thread is looping in [wti.c:165](https://github.com/rsyslog/rsyslog/blob/master/runtime/wti.c#L165) no other but the main thread was running though (seen in coredump).

I was not able to reproduce this with rsyslog running under GDB, so I resorted to [uprobes](http://www.brendangregg.com/blog/2015-06-28/linux-ftrace-uprobe.html).

Installed uprobes:

```
        'p:rsyslogd:wtiGetState %di +0x10(%di)' \
        'p:rsyslogd:wtiSetState %di %si +0x10(%di)' \
        'p:libpthread:pthread_create %dx %cx' \
        'p:libpthread:pthread_sigmask %di' \
        'p:libpthread:pthread_kill %di +0x2d0(%di)' \
        'p:libpthread:pthread_cancel %di +0x2d0(%di)' \
        'p:libpthread:_pthread_cleanup_pop %di' \
        'p:libpthread:sigcancel_handler %di' \
        'p:libpthread:pthread_setcancelstate %di' \
```

```
         rsyslogd-31608 [000] d... 4398887.145646: wtiSetState: (0x445a80) arg1=0x2115980 arg2=0x1 arg3=0x0
         rsyslogd-31608 [000] d... 4398887.145649: pthread_create: (0x7f01be2fba00) arg1=0x444e20 arg2=0x2115980 // thread 31761 creation
         
         rsyslogd-31608 [000] d... 4398887.145674: wtiGetState: (0x445a60) arg1=0x20fdec0 arg2=0x7f01b914b700
         rsyslogd-31608 [000] d... 4398887.145678: pthread_kill: (0x7f01be3010d0) arg1=0x7f01b914b700 arg2=0x7b7800007b80
  rs:action 1 que-31616 [000] d... 4398887.145861: pthread_setcancelstate: (0x7f01be3012d0) arg1=0x0
  rs:action 1 que-31616 [000] d... 4398887.145873: wtiSetState: (0x445a80) arg1=0x20fdec0 arg2=0x0 arg3=0x7f01b914b700
         rsyslogd-31608 [000] d... 4398887.146022: wtiGetState: (0x445a60) arg1=0x20fdec0 arg2=0x7f01b914b700
         rsyslogd-31608 [000] d... 4398887.146047: wtiGetState: (0x445a60) arg1=0x2115980 arg2=0x7f01ba94e700
         rsyslogd-31608 [000] d... 4398887.146052: pthread_kill: (0x7f01be3010d0) arg1=0x7f01ba94e700 arg2=0x7b7800007c11 //SIGTTIN
         
         rsyslogd-31608 [000] d... 4398887.146109: wtiGetState: (0x445a60) arg1=0x2115980 arg2=0x7f01ba94e700
         rsyslogd-31608 [000] d... 4398887.146114: pthread_kill: (0x7f01be3010d0) arg1=0x7f01ba94e700 arg2=0x7b7800007c11 //SIGTTIN
         
         rsyslogd-31608 [000] d... 4398887.146130: wtiGetState: (0x445a60) arg1=0x20fdec0 arg2=0x7f01b914b700
         rsyslogd-31608 [000] d... 4398887.146134: wtiGetState: (0x445a60) arg1=0x20fdec0 arg2=0x7f01b914b700
         rsyslogd-31608 [000] d... 4398887.146141: wtiGetState: (0x445a60) arg1=0x2115980 arg2=0x7f01ba94e700
         rsyslogd-31608 [000] d... 4398887.146145: pthread_kill: (0x7f01be3010d0) arg1=0x7f01ba94e700 arg2=0x7b7800007c11 //SIGTTIN
         
         rsyslogd-31608 [000] d... 4398887.156217: wtiGetState: (0x445a60) arg1=0x2115980 arg2=0x7f01ba94e700
         rsyslogd-31608 [000] d... 4398887.156247: pthread_cancel: (0x7f01be3011e0) arg1=0x7f01ba94e700 arg2=0x7b7800007c11 // 31761
         
         rsyslogd-31608 [000] d... 4398887.156252: wtiGetState: (0x445a60) arg1=0x2115980 arg2=0x7f01ba94e700
         rsyslogd-31760 [000] d... 4398887.157971: pthread_sigmask: (0x7f01be301060) arg1=0x0
```

During ~12 ms of the log above, the new thread is created and a signal (SIGTTIN and SIGCANCEL) are attempted to deliver to it.

If the thread 31761 was actually run, we would see it call [pthread_sigmask](https://github.com/rsyslog/rsyslog/blob/master/runtime/wtp.c#L388) [1] (we can see a previous thread 31760 manages to call that later [2]). The thread is thus terminated early at the beginning before it can do any setup (yet the running state is [already set](https://github.com/rsyslog/rsyslog/blob/master/runtime/wtp.c#L466)). 
FTR, from debug prints I found out that the "lost" thread is `queue[DA]:Reg`.

[1] I couldn't put a uprobe on `wtpWorker` due to symbol presence. 
[2] That shows that increased load prolonged intervals before a created thread is actually executed.

Cc: @nirmoy

**Comment by rgerhards** [comment]
Thanks for the bug report. Unfortunately, 8.4 is over 2 years old and heavily outdated. I think there is a big chance the the current version (8.30.0) does not have the issue. Can you update to it and see if the problem persists?

**Comment by rgerhards** [comment]
anyhow, this looks like a long-standing issue and it could be present in v8.30 as well. No easy repro in my test environment, but code review brings up some dubious things.

**Comment by rgerhards** [comment]
I seem to be able to at least get some reports from valgrind's helgrind thread debugger.

**Comment by Werkov** [comment]
Thanks for looking at this, reproducing this with newer rsyslog is still on by list -- it may happen that different threads are created during shutdown under that configuration and this race won't be manifested.

But I agree the thread pool implementation didn't change in this context. What do you think about introducing a third thread state ("starting")?

**Comment by rgerhards** [comment]
> But I agree the thread pool implementation didn't change in this context. What do you think about introducing a third thread state ("starting")?

lol, I just did that - among other things. See PR (it's preliminary and I expect that it changes a bit)

**Comment by Werkov** [comment]
So...
I tried the reproducer with a newer version of rsyslog, namely

> kernel-default-4.4.82-6.3.1.x86_64
> glibc-2.22-61.3.x86_64
> rsyslog-8.24.0-1.20.x86_64

And I wasn't able to reproduce it.

Then I tried to backport 6c332e20e812fd5d0c0d313ed90bdcfb17c4887d to 8.4.0 and when I was doing that, I ran into fae3e420f4e1da2b60400f4f06fdf41624e5fdf9. I backported only that single patch and I wasn't able to reproduce the original issue.

The explanation (I didn't trace the code) is that the spawning thread waits for proper setup of the worker child and thus it can't call pthread_cancel on an uninitialized thread.

The patch you posted (with third state) would make this more robust in case pthread_cancel was called from yet another thread (i.e. neither spawner nor spawnee).

Looking at TIDs in [uprobes trace](https://github.com/rsyslog/rsyslog/issues/1959#issue-270727662) we can see it is the parent in this instance who attempts the cancellation.
